packages/deployments/contracts/contracts/messaging/libraries/Queue.sol

```diff
@@ -106,8 +109,15 @@ library QueueLib {
     uint256 index; // Cursor for index in the batch of `items`.
     // NOTE: `first <= last` rephrased here to `!(first > last)` as it's a cheaper condition.
     while (!(first > last)) {
-      items[index] = queue.data[first];
+      bytes32 item = queue.data[first];
+      // Check to see if the item has been removed before appending it to the array.
+      if (!queue.removed[item]) {
+        items[index] = item;
+      }
```

Comment:
If item is removed then items[index] will not be updated. But in this case also index will get updated `++index; at L125`

This raises 2 problems:

1. This means returned items will have 0 address roots in middle of the indexes (if some root were blacklisted)
2. Since items size was defined beforehand `last + 1 - first`, it will always contain 0 address at end indexes if any of the root was blacklisted

Reply:
We could skip over an index in this event, and essentially move all the invalid items to the end of the array. We could track the count of items that were invalid, and then copy over into a new array the `<length of the old array> - <invalid item count>`, but that's not a very clean operation